### PR TITLE
Supporting repeating of tasks via -repeat

### PR DESCRIPTION
**Pull request by sergiocazzolato**
This commit introduces a new mechanism to reexecute jobs reusing the
context.
This commit includes 2 new options:
-retries <int>: defines the number of times to retry any task to be
executed
-stop-on-fail: indicates if the retries have to be stopped once the task
fails.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -516,6 +517,25 @@ as a `debug-each` script at the project, backend, and suite levels, so they
...
 
+<a name="reexecuting"/>
+
+## Reexecuting
+
+Reproduce an error could be a very boring experience, and Spread has a way to
+simplify that process by reexecuting the tasks as many times is desired until
+the task fails.
+
+To do that there are two options `-retries`, and `-stop-on-fail` which are
```
Comment (niemeyer) [inline README.md]: Retrying seems to have the meaning of _trying again if it fails_ but what we mean here is _run again until it fails_, so something like _-repeat 5_ would be more clear.

I also suggest getting rid of _-stop-on-fail_ and making that the default behavior. We don't want to watch the same task break 5 different times, perhaps in different ways. Every single failure is important and should be addressed.

File: cmd/spread/main.go [hunk 2]
```diff
@@ -32,6 +32,8 @@ var (
 	discard     = flag.Bool("discard", false, "Discard reused servers without running")
 	residue     = flag.String("residue", "", "Where to store residual data from tasks")
 	seed        = flag.Int64("seed", 0, "Seed for job order permutation")
+	retries     = flag.Int("retries", 0, "The number of reexecutions for each task")
```
Comment (niemeyer) [inline cmd/spread/main.go]:
```
repeat = flag.Int("repeat", 0, "Number of times to repeat each task")

File: README.md [hunk 3]
```diff
@@ -516,6 +517,25 @@ as a `debug-each` script at the project, backend, and suite levels, so they
 are aggregated and repeated for every task under them.
 
 
+<a name="reexecuting"/>
+
+## Reexecuting
```
Comment (niemeyer) [inline README.md]:
```
## Repeating runs

File: README.md [hunk 4]
```diff
@@ -516,6 +517,25 @@ as a `debug-each` script at the project, backend, and suite levels, so they
...
+
+Reproduce an error could be a very boring experience, and Spread has a way to
```
Comment (niemeyer) [inline README.md]: Reproducing an error may be

File: README.md [hunk 5]
```diff
@@ -516,6 +517,25 @@ as a `debug-each` script at the project, backend, and suite levels, so they
...
+simplify that process by reexecuting the tasks as many times is desired until
```
Comment (niemeyer) [inline README.md]: s/is/as/

File: README.md [hunk 6]
```diff
@@ -516,6 +517,25 @@ as a `debug-each` script at the project, backend, and suite levels, so they
...
+the task fails.
+
+To do that there are two options `-retries`, and `-stop-on-fail` which are
+used to determine the number of time to reexecute a task and if those executions
+have to be stoped when the task has failed.
+
+The `-retries` option receives an integer which indicates the number of retries 
+to do, being 0 the default value, and -1 can be used to reexecute indefinitely.
```
Comment (niemeyer) [inline README.md]: Let's please drop the -1 special behavior and that paragraph. People can easily type 100 or 10000.

File: README.md [hunk 7]
```diff
@@ -516,6 +517,25 @@ as a `debug-each` script at the project, backend, and suite levels, so they
...
+
+All the tasks listed to be executed will be reexercuted following the same
```
Comment (niemeyer) [inline README.md]: That may be dropped as well. It's implied that they're supposed to run similarly.

File: spread/runner.go [hunk 8]
```diff
@@ -594,27 +596,35 @@ func (r *Runner) worker(backend *Backend, system *System, order []int) {
...
-			r.add(&stats.TaskError, job)
-			debug = ""
-		}
-		if !abend && !r.options.Restore {
-			if err := r.fetchResidue(client, job); err != nil {
-				printf("Cannot fetch residue of %s: %v", job, err)
-				r.tomb.Killf("cannot fetch residue of %s: %v", job, err)
+		var retry = 0
+		var stopJob = false
+		for retry <= r.options.Retries && !stopJob {
```
Comment (niemeyer) [inline spread/runner.go]:
```
for repeat := r.options.Repeat; repeat >= 0; repeat-- {

File: spread/runner.go [hunk 9]
```diff
@@ -594,27 +596,35 @@ func (r *Runner) worker(backend *Backend, system *System, order []int) {
...
+			if r.options.Restore {
+				// Do not prepare or execute.
+			} else if !r.options.Restore && !r.run(client, job, preparing, job, job.Prepare(), debug, &abend) {
+				r.add(&stats.TaskPrepareError, job)
+				r.add(&stats.TaskAbort, job)
+				debug = ""
+				stopJob = r.options.StopOnFail
```
Comment (niemeyer) [inline spread/runner.go]:
```
repeat = -1
```
Same below.

File: README.md (hunk 3 above)
Comment (niemeyer) [inline README.md]: Updated to "tasks" instead of "runs".

File: spread/runner.go (hunk 9 above)
Comment (sergiocazzolato) [inline spread/runner.go]: Well in this case I need also to consider the scenario when I have to continue executing

Comment (niemeyer) [inline spread/runner.go]: Please see the earlier comment at the top. We don't want to repeat failing executions.

Comment (sergiocazzolato) [inline spread/runner.go]: ok, make sense

File: README.md [hunk 10]
```diff
@@ -516,6 +517,18 @@ as a `debug-each` script at the project, backend, and suite levels, so they
 are aggregated and repeated for every task under them.
 
 
+<a name="reexecuting"/>
```
Comment (niemeyer) [inline README.md]: repeating

File: README.md [hunk 11]
```diff
@@ -516,6 +517,18 @@ as a `debug-each` script at the project, backend, and suite levels, so they
...
+
+## Repeating runs
```
Comment (niemeyer) [inline README.md]: s/runs/tasks/ (sorry, I updated the comment but was probably too late)

File: cmd/spread/main.go [hunk 12]
```diff
@@ -32,6 +32,7 @@ var (
 	discard     = flag.Bool("discard", false, "Discard reused servers without running")
 	residue     = flag.String("residue", "", "Where to store residual data from tasks")
 	seed        = flag.Int64("seed", 0, "Seed for job order permutation")
+	repeat      = flag.Int("repeat", 0, "The number of reexecutions for each task")
```
Comment (niemeyer) [inline cmd/spread/main.go]: Per earlier comment:
```
repeat = flag.Int("repeat", 0, "Number of times to repeat each task")

File: spread/runner.go [hunk 13]
```diff
@@ -594,27 +595,32 @@ func (r *Runner) worker(backend *Backend, system *System, order []int) {
...
+				debug = ""
+				repeat = -1
+			} else if !r.options.Restore && r.run(client, job, executing, job, job.Task.Execute, debug, &abend) {
+				r.add(&stats.TaskDone, job)
+			} else if !r.options.Restore {
+				r.add(&stats.TaskError, job)
+				debug = ""
+				repeat = -1
+			}
+			if !abend && !r.options.Restore {
```
Comment (niemeyer) [inline spread/runner.go]:
```
... && repeat == 0
```
We don't want to fetch the same files every time.

File: spread/runner.go [hunk 14]
```diff
@@ -594,27 +595,32 @@ func (r *Runner) worker(backend *Backend, system *System, order []int) {
...
-			r.add(&stats.TaskError, job)
-			debug = ""
-		}
-		if !abend && !r.options.Restore {
-			if err := r.fetchResidue(client, job); err != nil {
-				printf("Cannot fetch residue of %s: %v", job, err)
-				r.tomb.Killf("cannot fetch residue of %s: %v", job, err)
+		for repeat := r.options.Repeat; repeat >= 0; repeat-- {
+			if r.options.Restore {
+				// Do not prepare or execute.
```
Comment (niemeyer) [inline spread/runner.go]:
```
// Do not prepare or execute, and don't repeat.
repeat = -1
```
We might conflict the options at some point, but for now it at least doesn't make sense to repeat the restore phase alone.

**Pull request merged**